src/inc/til/enumset.h

```diff
@@ -3,130 +3,134 @@
-        reference operator[](const Type pos)
-        {
-            return _base::operator[](static_cast<size_t>(pos));
+            return _data;
         }
 
         // Method Description:
         // - Returns the value of the bit at the given position.
         //   Throws std::out_of_range if it is not a valid position
         //   in the bitset.
-        bool test(const Type pos) const
+        bool test(const T pos) const noexcept
         {
-            return _base::test(static_cast<size_t>(pos));
+            const auto mask = to_underlying(pos);
+            return (_data & mask) != 0;
         }
 
         // Method Description:
         // - Returns true if any of the bits are set to true.
         bool any() const noexcept
         {
-            return _base::any();
+            return _data != 0;
         }
 
         // Method Description:
         // - Returns true if any of the bits in the given positions are true.
-        template<typename... Args, typename = std::enable_if_t<std::conjunction_v<std::is_same<Type, Args>...>>>
-        bool any(const Args... positions) const noexcept
+        TIL_ENUMSET_VARARG
+        bool any(Args... positions) const noexcept
         {
-            return (enumset{ positions... } & *this) != 0;
+            const auto mask = to_underlying(positions...);
+            return (_data & mask) != 0;
         }
 
         // Method Description:
         // - Returns true if all of the bits are set to true.
         bool all() const noexcept
         {
-            return _base::all();
+            return _data == ~underlying_type(0);
         }
 
         // Method Description:
         // - Returns true if all of the bits in the given positions are true.
-        template<typename... Args, typename = std::enable_if_t<std::conjunction_v<std::is_same<Type, Args>...>>>
-        bool all(const Args... positions) const noexcept
+        TIL_ENUMSET_VARARG
+        bool all(Args... positions) const noexcept
+        {
+            const auto mask = to_underlying(positions...);
+            return (_data & mask) == mask;
+        }
+
+        // Method Description:
+        // - Sets all of the bits in the given positions to true.
+        TIL_ENUMSET_VARARG
+        constexpr enumset& set(Args... positions) noexcept
         {
-            return (enumset{ positions... } & *this) == enumset{ positions... };
+            _data |= to_underlying(positions...);
+            return *this;
         }
 
         // Method Description:
         // - Sets the bit in the given position to the specified value.
-        enumset& set(const Type pos, const bool val = true)
+        constexpr enumset& set(const T pos, const bool val) noexcept
         {
-            _base::set(static_cast<size_t>(pos), val);
+            if (val)
+            {
+                set(pos);
+            }
+            else
+            {
+                reset(pos);
+            }

```

Comment:
Would it be worthwhile trying to avoid the branch here? Something like:
```
reset(pos);
_data |= (underlying_type{ val } << static_cast<underlying_type>(pos));
```

Reply:
I've did some benchmarks and settled on basically
```cpp
const auto mask = 1 << pos;
_data = (_data & ~mask) | (-val & mask);
```

Reply:
Even better. I love this.

Reply:
sh*t, that's clever